4626: clarify behavior (#4777)
ethereum/EIPs: EIPS/eip-4626.md

Tokenized Vaults have a lack of standardization leading to diverse implementation details.
Some various examples include lending markets (Compound, Aave, Fuse),
aggregators (Yearn, Rari Vaults, Idle), and intrinsically interest bearing tokens (xSushi).

> This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards and
> forces each protocol to implement their own adapters which are error prone and waste development resources.

Reviewer: ```suggestion
This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards,
and forces each protocol to implement their own adapters which are error prone and waste development resources.
```

ERC-4626 MAY implement [EIP-2612](./eip-2612.md) to improve the UX of approving shares on various integrations.

### Definitions:

- asset: The underlying token managed by the Vault.
  Has units defined by the corresponding ERC20 contract.

- share: The token of the Vault. Has a ratio of underlying assets
  exchanged on mint/deposit/withdraw/redeem (defined by the Vault).

- slippage: Any difference between advertised share price and economic realities of
  deposit to or withdrawal from the Vault, which is not accounted by fees.

- fee: An amount of assets or shares charged to the Vault. Can be a fee on deposits,

>   yield, AUM, withdrawal, or any other kind of fee.

Reviewer: Besides the extra spacing between the defs, I don't think this indentation changes anything in how it's displayed

  inputs: []

  outputs:
    - name: totalAssets
      type: uint256
```

#### assetsPerShare

The current exchange rate of shares to assets, quoted per share unit (a share unit is `10 ** Vault.decimals()`).

MUST be equal to `(Vault.totalAssets() * (10 ** Vault.decimals())) / Vault.totalSupply()` unless `Vault.totalSupply()` is 0.

MUST be inclusive of any fees that are charged against assets in the Vault.

> MAY _NOT_ be completely accurate according to slippage or other on-chain conditions, when performing the actual exchange.

Reviewer: I think should add the note about manipulation resistance

> MUST _NOT_ return 0 to indicate the Vault is empty.

Reviewer: ```suggestion
MUST _NOT_ return 0 the Vault has no assets.
```

but potentially change to:
```
SHOULD return a 1:1 exchange rate when the Vault is empty
```

Author: the suggestion makes no sense read it back lol

Author: > SHOULD return a 1:1 exchange rate when the Vault is empty

what if ppl want 2:1 share ratios n such? idt worth restricting

Reviewer: > > SHOULD return a 1:1 exchange rate when the Vault is empty
> 
> what if ppl want 2:1 share ratios n such? idt worth restricting

I agree. How about "MUST return the initial share price when empty"?

Reviewer: ```suggestion
MUST return the initial share price when empty.
```

#### assetsOf

Total number of underlying assets that `depositor`'s shares represent.

MAY be more accurate than using `assetsPerShare` or `totalAssets / Vault.totalSupply` for certain types of fee calculations.

MAY _NOT_ be completely accurate according to slippage or other on-chain conditions.

MUST _NOT_ revert.

```yaml
- name: assetsOf
  type: function
  stateMutability: view

  inputs:
    - name: depositor
      type: address

  outputs:
    - name: assets
      type: uint256
```

#### maxDeposit

Total number of underlying assets that `caller` can be deposit.

MUST return a limited value if `caller` is subject to some deposit limit.

MUST return `2 ** 256 - 1` if there is no limit on the maximum amount of assets that may be deposited.

MAY be used in the `previewDeposit` or `deposit` methods for `assets` input parameter.

MUST _NOT_ revert.

```yaml
- name: maxDeposit
  type: function
  stateMutability: view

  inputs:
    - name: caller
      type: address

  outputs:
    - name: maxAssets
      type: uint256
```

#### previewDeposit

Allows an on-chain or off-chain user to simulate the effects of their deposit at the current block,
given current on-chain conditions.

> MUST return as close to the _exact_ amount of Vault shares that would be minted if the caller were

Reviewer: Does it under shoot or overshoot, or is it within some bound of what the correct answer might be? What properties are different versus `assetsPerShare`?

> to deposit a given _exact_ amount of underlying assets using the `deposit` method as technically possible.

Reviewer 2: I think thit is still not explicitly stating that balance check cannot be made.

> MUST return as close to the _exact_ amount of Vault shares that would be minted if the caller were
> to deposit a given _exact_ amount of underlying assets using the `deposit` method as technically possible.

Reviewer: Should be one of the following:
```suggestion
MUST return the _minimum_ amount of Vault shares that would be minted if the caller were
to deposit a given _exact_ amount of underlying assets using the `deposit` method.
```

Reviewer: ```suggestion
MUST return the _maximum_ amount of Vault shares that would be minted if the caller were
to deposit a given _exact_ amount of underlying assets using the `deposit` method.
```

Reviewer: ```suggestion
MUST return an amount of Vault shares that would be minted if the caller were to deposit a given
 _exact_ amount of underlying assets using the `deposit` method, within an `X` bps range of accuracy.
```
where `X` is something like 1 bps (basis point e.g. 0.01%)

Author: idt its realistic to enforce a numerical bound here

Reviewer: So how about "MUST return some number between 0 and `2 ** 256 - 1` that represents an approximation of the exchange rate for the exact number of shares (but really it's just a random number lol, don't ask for specifics)"?

MUST _NOT_ account for deposit limits like those returned from `maxDeposit` and should always act

> as though the deposit would be accepted, regardless if the user has enough tokens approved, etc.

Reviewer: Make a note how `max*` accounts for this instead?

MUST be inclusive of deposit fees. Integrators should be aware of the existence of deposit fees.

MUST _NOT_ revert.

> Note that any unfavorable discrepancy between `assetsPerShare` and the ratio of `assets` to `shares` SHOULD

Reviewer: This is where having the exchange function directly would be helpful

Reviewer 3: 🤩

Author: exactly :)
